index/scorch/snapshot_index.go

```diff
@@ -507,11 +523,29 @@ func (i *IndexSnapshot) documentVisitFieldTerms(id index.IndexInternalID,
 	ss := i.segment[segmentIndex]
 
 	if zaps, ok := ss.segment.(segment.DocumentFieldTermVisitable); ok {
-		// get the list of doc value persisted fields
-		pFields, err := zaps.VisitableDocValueFields()
-		if err != nil {
-			return nil, err
+		pFields := ss.cachedDocs.getPersistedFields()
+		// if all fields are dv presisted
+		if len(pFields) > 0 && checkSubset(pFields, fields) {
+			return zaps.VisitDocumentFieldTerms(localDocNum, fields, visitor, dvs)
+		}
+
+		cFields := ss.cachedDocs.getCachedFields()
+		// if all fields are already in dvCache or persisted
+		if len(cFields) > 0 && checkSubset(append(pFields, cFields...), fields) {
+			dvPendingFields := extractDvPendingFields(fields, pFields)
+			visitDocumentFieldCacheTerms(localDocNum, dvPendingFields, ss, visitor)
```

Comment:
looks like visitDocumentFieldCacheTerms() is not multithread safe, as it accesses the shared cache without acquiring the lock.  It should likely instead be a method of the cachedDocs struct, as it hardly uses the ss param.

Reply:
The repetition of the visitDocumentFieldCacheTerms() and VisitDocumentFieldTerms() and multiple return exit conditions doesn't have a cleanliness to it.  Instead, I find it easier to think about if there were no early exits and "all the waters have one exit in the Nile delta".

Reply:
so again, this was already discussed, sreekanth did not make any attempt at that clean up in this change because he spent all day at the passport office.  we should sync up to discuss further.

Reply:
one thought is,  "if there were no early exits " and if we end up moving the visitDocumentFieldCacheTerms() under cachedDocs within a mutex, then we also need ensure that we don't call it when not needed(when not served), otherwise it will add a lock in a hot path...